pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,242 @@
+	// `IndexEntries`).
+	keyAsRowCount := roachpb.RowCount{
+		DataSize: int64(len(key.Key)) + int64(len(value)),
+	}
+
+	if err := s.sst.PutRawMVCC(elidedKey, value); err != nil {
+		return err
+	}
+
+	lastFile.EntryCounts.Add(keyAsRowCount)
+	s.flushedSize += keyAsRowCount.DataSize
+	s.prevKey = append(s.prevKey[:0], key.Key...)
+
+	return nil
+}
+
+// Reset resets the SSTSinkKeyWriter to write to a new span. It ensures that an
+// open BackupManifest_File exists for the new span. It will either open a new
+// BackupManifest_File, or if the new span contiguously extends the last span,
+// it will extend it if it remains under the fileSpanByteLimit. The caller is
+// responsible for ensuring that spans that are reset on do not end mid-row.
+// Any time a new span is being written, Reset MUST be called prior to any
+// WriteKey calls.
+func (s *SSTSinkKeyWriter) Reset(ctx context.Context, newSpan roachpb.Span) error {
+	log.VEventf(ctx, 2, "resetting sink to span %s", newSpan)
+	if err := s.maybeFlushNonExtendableSpans(newSpan); err != nil {
+		return err
+	}
+	if s.out == nil {
+		if err := s.open(ctx); err != nil {
+			return err
+		}
+	}
+	// At this point, we can assume the new span is either contiguous with the
+	// last span or comes after the last span (if it exists).
+	var lastFile *backuppb.BackupManifest_File
+	if len(s.flushedFiles) > 0 {
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	if lastFile != nil {
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+		if isContiguousSpan(lastFile.Span, newSpan) &&
+			(s.midRow || lastFile.EntryCounts.DataSize < fileSpanByteLimit) {
+			log.VEventf(ctx, 2, "extending span %s to %s", lastFile.Span, newSpan)
+			s.stats.spanGrows++
+			lastFile.Span.EndKey = newSpan.EndKey
+			return nil
+		}
+	}
+
+	s.flushedFiles = append(
+		s.flushedFiles,
+		backuppb.BackupManifest_File{
+			Span: newSpan,
+			Path: s.outName,
+		},
+	)
+	s.prevKey = nil
+	return nil
+}
+
+func (s *SSTSinkKeyWriter) Flush(ctx context.Context) error {
+	if len(s.flushedFiles) > 0 {
+		lastFile := s.flushedFiles[len(s.flushedFiles)-1]
+		// If WriteKey was used, it is possible that the last written key was
+		// mid-row, but because no keys were written after, s.midRow was not updated.
+		// To ensure this, we update midRow using the end of the span that the last
+		// key belonged to.
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+	}
+	err := s.FileSSTSink.Flush(ctx)
+	return err
+}
+
+// maybeFlushNonExtendableSpans checks if the new span's startKey precedes the
+// last spans' endKey or if the two spans do not share the same prefix based on
+// the sink's ElideMode. If either of these conditions are met, the current
+// backup file is flushed and a new one is created.
+func (s *SSTSinkKeyWriter) maybeFlushNonExtendableSpans(newSpan roachpb.Span) error {
+	if len(s.flushedFiles) == 0 {
+		return nil
+	}
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
```

Comment:
why are you instantiating a pointer to a file? In general, it's best practice to create a local copy of a var to avoid accidental in place modifications.